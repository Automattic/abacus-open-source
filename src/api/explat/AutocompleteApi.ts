import { StatusCodes } from 'http-status-codes'

import { fetchApi } from 'src/api/explat/utils'
import { AutocompleteItem, autocompleteSchema, eventDetailsSchema } from 'src/lib/explat/schemas'

import HttpResponseError from '../HttpResponseError'

async function getCompletion(name: string) {
  return await autocompleteSchema.validate(await fetchApi('GET', `/autocomplete/${name}`), { abortEarly: false })
}

export async function getUserCompletions(): Promise<AutocompleteItem[]> {
  return (await getCompletion('users')).completions
}

export async function getEventNameCompletions(): Promise<AutocompleteItem[]> {
  return (await getCompletion('events')).completions
}

export async function getPropNameCompletions(eventName: string): Promise<AutocompleteItem[] | null> {
  if (!eventName) {
    throw new Error('No eventName to getPropNameCompletions of.')
  }

  try {
    const apiResponse = await eventDetailsSchema.validate(await fetchApi('GET', `/autocomplete/events/${eventName}`))
    return apiResponse.props.map((p) => ({
      name: p.name,
      value: p.name,
    }))
  } catch (error) {
    // istanbul ignore else; Forced to use the else here to prevent two istanbul ignores
    if (error instanceof HttpResponseError && error.status === StatusCodes.NOT_FOUND) {
      return null
    } else {
      throw error
    }
  }
}
