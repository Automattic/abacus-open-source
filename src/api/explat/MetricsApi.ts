import _ from 'lodash'
import * as yup from 'yup'

import {
  Metric,
  MetricNew,
  MetricNewOutbound,
  metricNewOutboundSchema,
  metricNewSchema,
  metricSchema,
} from 'src/lib/explat/schemas'

import { fetchApi } from './utils'

/**
 * Transforms a metric into an outbound form.
 */

function outboundTransformMetric(metricNew: MetricNew): MetricNewOutbound {
  const metricDataForRequest = {
    ...metricNew,
    tags: metricNew.tags.map((tagId) => ({ tagId })),
  }
  return metricNewOutboundSchema.cast(metricDataForRequest)
}

/**
 * Attempts to create a new metric.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function create(newMetric: MetricNew): Promise<Metric> {
  const validatedNewMetric = await metricNewSchema.validate(newMetric, { abortEarly: false })
  const outboundNewMetric = outboundTransformMetric(validatedNewMetric)
  return await metricSchema.validate(await fetchApi('POST', '/metrics', outboundNewMetric))
}

/**
 * Attempts to put a new metric.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function put(metricId: number, newMetric: MetricNew): Promise<Metric> {
  // istanbul ignore next; Shouldn't happen
  if (!_.isNumber(metricId)) {
    throw new Error('Invalid metricId.')
  }
  const validatedNewMetric = await metricNewSchema.validate(newMetric, { abortEarly: false })
  const outboundNewMetric = outboundTransformMetric(validatedNewMetric)
  return await metricSchema.validate(await fetchApi('PUT', `/metrics/${metricId}`, outboundNewMetric))
}

/**
 * Finds all the available metrics.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @throws UnauthorizedError
 */
async function findAll(options: { includeDebug: boolean } = { includeDebug: false }): Promise<Metric[]> {
  // istanbul ignore next; debug only
  const { metrics } = await yup
    .object({ metrics: yup.array(metricSchema).defined() })
    .defined()
    .validate(await fetchApi('GET', options.includeDebug ? '/metrics?debug=true' : '/metrics'), {
      abortEarly: false,
    })
  return metrics
}

/**
 * Find the metric by ID.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @throws UnauthorizedError
 */
async function findById(metricId: number): Promise<Metric> {
  return await metricSchema.validate(await fetchApi('GET', `/metrics/${metricId}`), { abortEarly: false })
}

const MetricsApi = {
  create,
  put,
  findAll,
  findById,
}

export default MetricsApi
