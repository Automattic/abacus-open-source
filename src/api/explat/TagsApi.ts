import _ from 'lodash'
import * as yup from 'yup'

import {
  TagBare,
  tagBareSchema,
  TagFull,
  TagFullNew,
  tagFullNewOutboundSchema,
  tagFullNewSchema,
  tagFullSchema,
} from 'src/lib/explat/schemas'
import { isDebugMode } from 'src/utils/general'

import { fetchApi } from './utils'

/**
 * Attempts to create a new tag.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function create(newTag: TagFullNew): Promise<TagBare> {
  const validatedNewTag = await tagFullNewSchema.validate(newTag, { abortEarly: false })
  const outboundNewTag = tagFullNewOutboundSchema.cast(validatedNewTag)
  return await tagBareSchema.validate(await fetchApi('POST', '/tags', outboundNewTag))
}

/**
 * Attempts to put a new tag.
 *
 * Note: Be sure to handle any errors that may be thrown.
 */
async function put(tagId: number, newTag: TagFullNew): Promise<TagBare> {
  // istanbul ignore next; Shouldn't happen
  if (!_.isNumber(tagId)) {
    throw new Error('Invalid tagId.')
  }
  const validatedNewTag = await tagFullNewSchema.validate(newTag, { abortEarly: false })
  const outboundNewTag = tagFullNewOutboundSchema.cast(validatedNewTag)
  return await tagBareSchema.validate(await fetchApi('PUT', `/tags/${tagId}`, outboundNewTag))
}

/**
 * Finds all the available tags.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @throws UnauthorizedError
 */
async function findAll(): Promise<TagBare[]> {
  // istanbul ignore next; debug only
  const { tags } = await yup
    .object({ tags: yup.array(tagBareSchema).defined() })
    .defined()
    .validate(await fetchApi('GET', isDebugMode() ? '/tags?debug=true' : '/tags'), {
      abortEarly: false,
    })
  return tags
}

/**
 * Find the tag by ID.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @throws UnauthorizedError
 */
async function findById(tagId: number): Promise<TagFull> {
  return await tagFullSchema.validate(await fetchApi('GET', `/tags/${tagId}`), { abortEarly: false })
}

const TagsApi = {
  create,
  put,
  findAll,
  findById,
}

export default TagsApi
