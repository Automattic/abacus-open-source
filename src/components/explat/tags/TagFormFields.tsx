import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import { Field } from 'formik'
import { TextField } from 'formik-material-ui'
import React from 'react'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    root: {},
    row: {
      margin: theme.spacing(3, 0),
      display: 'flex',
      alignItems: 'center',
      '&:first-of-type': {
        marginTop: 0,
      },
    },
  }),
)

const TagFormFields = (): JSX.Element => {
  const classes = useStyles()

  return (
    <>
      <div className={classes.row}>
        <Field
          component={TextField}
          name='tag.namespace'
          id='tag.namespace'
          label='Tag namespace'
          placeholder='tag_namespace'
          helperText='Use snake_case.'
          variant='outlined'
          fullWidth
          required
          InputLabelProps={{
            shrink: true,
          }}
        />
      </div>
      <div className={classes.row}>
        <Field
          component={TextField}
          name='tag.name'
          id='tag.name'
          label='Tag name'
          placeholder='tag_name'
          helperText='Use snake_case.'
          variant='outlined'
          fullWidth
          required
          InputLabelProps={{
            shrink: true,
          }}
        />
      </div>
      <div className={classes.row}>
        <Field
          component={TextField}
          name='tag.description'
          id='tag.description'
          label='Tag description'
          placeholder='Put your Tag description here!'
          variant='outlined'
          fullWidth
          required
          multiline
          rows={4}
          InputLabelProps={{
            shrink: true,
          }}
        />
      </div>
    </>
  )
}

export default TagFormFields
