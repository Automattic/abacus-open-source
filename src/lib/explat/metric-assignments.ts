import _ from 'lodash'

import { AttributionWindowSeconds, MetricAssignment } from './schemas'

export const AttributionWindowSecondsToHuman: Record<AttributionWindowSeconds, string> = {
  [AttributionWindowSeconds.OneHour]: '1 hour',
  [AttributionWindowSeconds.SixHours]: '6 hours',
  [AttributionWindowSeconds.TwelveHours]: '12 hours',
  [AttributionWindowSeconds.TwentyFourHours]: '24 hours',
  [AttributionWindowSeconds.SeventyTwoHours]: '72 hours',
  [AttributionWindowSeconds.OneWeek]: '1 week',
  [AttributionWindowSeconds.TwoWeeks]: '2 weeks',
  [AttributionWindowSeconds.ThreeWeeks]: '3 weeks',
  [AttributionWindowSeconds.FourWeeks]: '4 weeks',
}

/**
 * Return the experiment's metric assignments sorted in a canonical order:
 * - Primary first
 * - Assignments with the same metricId are next to each other
 * - Assignments with the same metricId are ordered by attributionWindow asc
 * -
 */
export function sort(metricAssignments: MetricAssignment[]): MetricAssignment[] {
  const metricAssignmentsByMetric = Object.values(_.groupBy(metricAssignments, 'metricId'))
    // Order within groups
    .map((metricAssignments) =>
      _.orderBy(metricAssignments, ['isPrimary', 'attributionWindowSeconds'], ['desc', 'asc']),
    )
  // Order the groups
  const orderedMetricAssignmentsByMetric = _.orderBy(
    metricAssignmentsByMetric,
    ['[0].isPrimary', '[0].metricId'],
    ['desc', 'asc'],
  )

  return _.flatten(orderedMetricAssignmentsByMetric)
}
