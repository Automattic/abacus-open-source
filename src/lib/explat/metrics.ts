import { match, P } from 'ts-pattern'

import { Metric, MetricParameterType, PipeModels, PipeValueFields } from './schemas'

/**
 * Return a stringified version of the metric params object
 */
export const stringifyMetricParams = (metric: Metric): string =>
  JSON.stringify(metric.parameterType === 'conversion' ? metric.eventParams : metric.revenueParams, null, 4)

export enum UnitType {
  Ratio = 'ratio',
  RatioPoints = 'ratio_points',
  Count = 'count',
  Usd = 'usd',
}

export const metricParameterTypeName: Record<MetricParameterType, string> = {
  [MetricParameterType.Revenue]: 'Cash Sales',
  [MetricParameterType.Conversion]: 'Conversion',
  [MetricParameterType.Pipe]: 'Pipe Model',
} as const

export interface UnitInfo {
  unitType: UnitType
}

export enum UnitDerivationType {
  AbsoluteDifference = 'AbsoluteDifference',
  RelativeDifference = 'RelativeDifference',
  ImpactScaled = 'ImpactScaled',
}

export function getUnitInfo(metric: Metric, derivations: UnitDerivationType[] = []): UnitInfo {
  // So the real way to handle units is to have better data on the units, e.g.
  // instead of using Conversion we use users-converted / users-sample, this
  // ends up being very complicated and not something made for Javascript.
  // We are now using the second best thing, taking a metric, and what we want
  // to do to it, and then determine a UnitType from that.
  //
  // For implementation simplicity we are using pattern matching so we don't
  // have a massive multi-level case statements:

  const conversionMetricUnion = P.union(
    { parameterType: MetricParameterType.Conversion },
    {
      parameterType: MetricParameterType.Pipe,
      pipeParams: { model: PipeModels.ChurnUntimed, valueField: PipeValueFields.Prediction },
    },
  )

  return match({ metric, derivations })
    .with(
      {
        metric: conversionMetricUnion,
        derivations: [],
      },
      () => ({ unitType: UnitType.Ratio }),
    )
    .with(
      {
        metric: conversionMetricUnion,
        derivations: [UnitDerivationType.AbsoluteDifference],
      },
      () => ({ unitType: UnitType.RatioPoints }),
    )
    .with(
      {
        metric: conversionMetricUnion,
        derivations: [UnitDerivationType.ImpactScaled],
      },
      {
        metric: conversionMetricUnion,
        derivations: [UnitDerivationType.AbsoluteDifference, UnitDerivationType.ImpactScaled],
      },
      () => ({ unitType: UnitType.Count }),
    )
    .with(
      {
        metric: { parameterType: MetricParameterType.Revenue },
        derivations: [],
      },
      () => ({ unitType: UnitType.Usd }),
    )
    .with(
      {
        metric: { parameterType: MetricParameterType.Revenue },
        derivations: [UnitDerivationType.AbsoluteDifference],
      },
      () => ({ unitType: UnitType.Usd }),
    )
    .with(
      {
        metric: { parameterType: MetricParameterType.Revenue },
        derivations: [UnitDerivationType.ImpactScaled],
      },
      {
        metric: { parameterType: MetricParameterType.Revenue },
        derivations: [UnitDerivationType.AbsoluteDifference, UnitDerivationType.ImpactScaled],
      },
      () => ({ unitType: UnitType.Usd }),
    )
    .with(
      {
        derivations: [UnitDerivationType.RelativeDifference],
      },
      () => ({ unitType: UnitType.Ratio }),
    )
    .otherwise(() => {
      throw new Error('Could not find matching unit-type for metric and derivations.')
    })
}
