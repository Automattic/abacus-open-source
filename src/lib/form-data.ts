import { formatIsoDate } from 'src/utils/time'

import {
  Event,
  ExperimentFull,
  Metric,
  MetricAssignment,
  MetricParameterType,
  SegmentAssignment,
  TagBare,
  Variation,
} from './schemas'

interface MetricAssignmentFormData {
  attributionWindowSeconds: string
  changeExpected: boolean
  metricId: string
  isPrimary: boolean
  minDifference: string
}

interface SegmentAssignmentFormData {
  isExcluded: boolean
  segmentId: number
}

interface VariationFormData {
  allocatedPercentage: string
  isDefault: boolean
  name: string
}

interface ExposureEventFormData {
  event: string
  props: { value: unknown; key: string }[]
}

function metricAssignmentToFormData(metricAssignment: MetricAssignment): MetricAssignmentFormData {
  return {
    metricId: String(metricAssignment.metricId),
    attributionWindowSeconds: String(metricAssignment.attributionWindowSeconds),
    isPrimary: metricAssignment.isPrimary,
    changeExpected: metricAssignment.changeExpected,
    minDifference: String(metricAssignment.minDifference),
  }
}

function segmentAssignmentToFormData(segmentAssignment: SegmentAssignment): SegmentAssignmentFormData {
  return {
    segmentId: segmentAssignment.segmentId,
    isExcluded: segmentAssignment.isExcluded,
  }
}

function variationToFormData(variation: Variation): VariationFormData {
  return {
    name: variation.name,
    isDefault: variation.isDefault,
    allocatedPercentage: String(variation.allocatedPercentage),
  }
}

function exposureEventToFormData(exposureEvent: Event): ExposureEventFormData {
  return {
    event: exposureEvent.event,
    props: Object.entries(exposureEvent.props || {}).map(([key, value]) => ({ key, value })),
  }
}

/**
 * Takes an experiment object and formats it for use as form-data in ExperimentForm.
 */
export function experimentToFormData(experiment: Partial<ExperimentFull>): {
  startDatetime: string
  variations: VariationFormData[]
  segmentAssignments: SegmentAssignmentFormData[]
  name: string
  description: string
  metricAssignments: MetricAssignmentFormData[]
  exposureEvents: ExposureEventFormData[]
  existingUsersAllowed: 'true' | 'false'
  endDatetime: string
  platform: string
  p2Url: string
  ownerLogin: string
  exclusionGroupTagIds?: number[]
  platformSegments: Record<string, string>
} {
  return {
    p2Url: experiment.p2Url ?? '',
    name: experiment.name ?? '',
    description: experiment.description ?? '',
    startDatetime: experiment.startDatetime ? formatIsoDate(experiment.startDatetime) : '',
    endDatetime: experiment.endDatetime ? formatIsoDate(experiment.endDatetime) : '',
    ownerLogin: experiment.ownerLogin ?? '',
    existingUsersAllowed:
      experiment.existingUsersAllowed === undefined
        ? 'true'
        : (String(experiment.existingUsersAllowed) as 'true' | 'false'),
    platform: experiment.platform ?? '',
    metricAssignments: experiment.metricAssignments ? experiment.metricAssignments.map(metricAssignmentToFormData) : [],
    segmentAssignments: experiment.segmentAssignments
      ? experiment.segmentAssignments.map(segmentAssignmentToFormData)
      : [],
    variations: experiment.variations
      ? experiment.variations.map(variationToFormData)
      : [
          { name: 'control', isDefault: true, allocatedPercentage: '50' },
          { name: 'treatment', isDefault: false, allocatedPercentage: '50' },
        ],
    exposureEvents: experiment.exposureEvents ? experiment.exposureEvents.map(exposureEventToFormData) : [],
    exclusionGroupTagIds: experiment.exclusionGroupTagIds ?? [],
    platformSegments: (experiment.platformSegments || []).reduce((segments, platformSegment) => {
      return {
        ...segments,
        [platformSegment.name]: platformSegment.value,
      }
    }, {}),
  }
}
export type ExperimentFormData = ReturnType<typeof experimentToFormData>

/**
 * Convert a metric for use as form data in Formik.
 */
export const metricToFormData: (metric: Partial<Metric>) => {
  parameterType: MetricParameterType
  name: string
  eventParams: string | undefined
  description: string
  revenueParams: string | undefined
  higherIsBetter: boolean
} = (metric: Partial<Metric>) => ({
  name: metric.name ?? '',
  description: metric.description ?? '',
  parameterType: metric.parameterType ?? MetricParameterType.Conversion,
  higherIsBetter: metric.higherIsBetter ?? true,
  eventParams: metric.eventParams ? JSON.stringify(metric.eventParams, null, 2) : undefined,
  revenueParams: metric.revenueParams ? JSON.stringify(metric.revenueParams, null, 2) : undefined,
})
export type MetricFormData = ReturnType<typeof metricToFormData>

/**
 * Convert a tag for use as form data in Formik.
 */
export const tagToFormData: (tag: Partial<TagBare>) => {
  name: string
  description: string
} = (tag: Partial<TagBare>) => ({
  namespace: tag.namespace ?? '',
  name: tag.name ?? '',
  description: tag.description ?? '',
})
export type TagFormData = ReturnType<typeof tagToFormData>
