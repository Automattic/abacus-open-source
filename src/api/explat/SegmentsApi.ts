import * as yup from 'yup'

import { Segment, segmentSchema } from 'src/lib/explat/schemas'

import { fetchApi } from './utils'

/**
 * Finds all the available segments.
 *
 * Note: Be sure to handle any errors that may be thrown.
 *
 * @throws UnauthorizedError
 */
async function findAll(): Promise<Segment[]> {
  const { segments } = await yup
    .object({ segments: yup.array(segmentSchema).defined() })
    .defined()
    .validate(await fetchApi('GET', '/segments'), { abortEarly: false })
  return segments
}

const SegmentsApi = {
  findAll,
}

export default SegmentsApi
