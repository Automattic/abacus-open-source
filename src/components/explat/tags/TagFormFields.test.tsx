/* eslint-disable @typescript-eslint/require-await */
import { Formik } from 'formik'
import MockDate from 'mockdate'
import React from 'react'

import { tagToFormData } from 'src/lib/explat/form-data'
import Fixtures from 'src/test-helpers/fixtures'
import { render } from 'src/test-helpers/test-utils'

import TagFormFields from './TagFormFields'

MockDate.set('2020-07-21')

test('renders as expected for an existing tag', async () => {
  const tag = Fixtures.createTagFull(1)
  const { container } = render(
    <Formik
      initialValues={{
        tag: tagToFormData(tag),
      }}
      onSubmit={
        /* istanbul ignore next; This is unused */
        () => undefined
      }
    >
      {() => <TagFormFields />}
    </Formik>,
  )

  expect(container).toMatchSnapshot()
})

test('renders as expected for a new tag', () => {
  const { container } = render(
    <Formik
      initialValues={{
        tag: tagToFormData({}),
      }}
      onSubmit={
        /* istanbul ignore next; This is unused */
        () => undefined
      }
    >
      {() => <TagFormFields />}
    </Formik>,
  )

  expect(container).toMatchSnapshot()
})
