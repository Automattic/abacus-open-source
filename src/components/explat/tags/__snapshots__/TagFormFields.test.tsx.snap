// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected for a new tag 1`] = `
<div>
  <div
    class="makeStyles-row-8"
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-required Mui-required"
        data-shrink="true"
        for="tag.namespace"
        id="tag.namespace-label"
      >
        Tag namespace
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
      >
        <input
          aria-describedby="tag.namespace-helper-text"
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          id="tag.namespace"
          name="tag.namespace"
          placeholder="tag_namespace"
          required=""
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-9 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-11 PrivateNotchedOutline-legendNotched-12"
          >
            <span>
              Tag namespace
               *
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained Mui-required"
        id="tag.namespace-helper-text"
      >
        Use snake_case.
      </p>
    </div>
  </div>
  <div
    class="makeStyles-row-8"
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-required Mui-required"
        data-shrink="true"
        for="tag.name"
        id="tag.name-label"
      >
        Tag name
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
      >
        <input
          aria-describedby="tag.name-helper-text"
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          id="tag.name"
          name="tag.name"
          placeholder="tag_name"
          required=""
          type="text"
          value=""
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-9 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-11 PrivateNotchedOutline-legendNotched-12"
          >
            <span>
              Tag name
               *
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained Mui-required"
        id="tag.name-helper-text"
      >
        Use snake_case.
      </p>
    </div>
  </div>
  <div
    class="makeStyles-row-8"
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined Mui-required Mui-required"
        data-shrink="true"
        for="tag.description"
        id="tag.description-label"
      >
        Tag description
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
      >
        <textarea
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
          id="tag.description"
          name="tag.description"
          placeholder="Put your Tag description here!"
          required=""
          rows="4"
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-9 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-11 PrivateNotchedOutline-legendNotched-12"
          >
            <span>
              Tag description
               *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;

exports[`renders as expected for an existing tag 1`] = `
<div>
  <div
    class="makeStyles-row-2"
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled Mui-required Mui-required"
        data-shrink="true"
        for="tag.namespace"
        id="tag.namespace-label"
      >
        Tag namespace
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
      >
        <input
          aria-describedby="tag.namespace-helper-text"
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          id="tag.namespace"
          name="tag.namespace"
          placeholder="tag_namespace"
          required=""
          type="text"
          value="tag_namespace_1"
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-5 PrivateNotchedOutline-legendNotched-6"
          >
            <span>
              Tag namespace
               *
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained MuiFormHelperText-filled Mui-required"
        id="tag.namespace-helper-text"
      >
        Use snake_case.
      </p>
    </div>
  </div>
  <div
    class="makeStyles-row-2"
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled Mui-required Mui-required"
        data-shrink="true"
        for="tag.name"
        id="tag.name-label"
      >
        Tag name
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
      >
        <input
          aria-describedby="tag.name-helper-text"
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input"
          id="tag.name"
          name="tag.name"
          placeholder="tag_name"
          required=""
          type="text"
          value="tag_1"
        />
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-5 PrivateNotchedOutline-legendNotched-6"
          >
            <span>
              Tag name
               *
            </span>
          </legend>
        </fieldset>
      </div>
      <p
        class="MuiFormHelperText-root MuiFormHelperText-contained MuiFormHelperText-filled Mui-required"
        id="tag.name-helper-text"
      >
        Use snake_case.
      </p>
    </div>
  </div>
  <div
    class="makeStyles-row-2"
  >
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
    >
      <label
        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-filled Mui-required Mui-required"
        data-shrink="true"
        for="tag.description"
        id="tag.description-label"
      >
        Tag description
        <span
          aria-hidden="true"
          class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
        >
           
          *
        </span>
      </label>
      <div
        class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiOutlinedInput-multiline"
      >
        <textarea
          aria-invalid="false"
          class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline MuiOutlinedInput-inputMultiline"
          id="tag.description"
          name="tag.description"
          placeholder="Put your Tag description here!"
          required=""
          rows="4"
        >
          This is tag 1
        </textarea>
        <fieldset
          aria-hidden="true"
          class="PrivateNotchedOutline-root-3 MuiOutlinedInput-notchedOutline"
        >
          <legend
            class="PrivateNotchedOutline-legendLabelled-5 PrivateNotchedOutline-legendNotched-6"
          >
            <span>
              Tag description
               *
            </span>
          </legend>
        </fieldset>
      </div>
    </div>
  </div>
</div>
`;
